File: quoridor/training/mcts.py
"""PUCT MCTS guided by a policy/value network."""

from __future__ import annotations

import math

import numpy as np

from ..engine.game import (
    ACTION_SIZE,
    apply_action,
    encode_action_for_player,
    encode_state,
    legal_action_mask,
)
from ..engine.state import State

C_PUCT = 1.5


class Node:
    __slots__ = ("state", "prior", "children", "visit_count", "value_sum", "player")

    def __init__(self, state: State, prior: float = 0.0):
        self.state = state
        self.prior = prior
        self.children: dict[int, "Node"] = {}
        self.visit_count = 0
        self.value_sum = 0.0
        self.player = state.current_player

    def value(self) -> float:
        return self.value_sum / self.visit_count if self.visit_count else 0.0

    def expanded(self) -> bool:
        return len(self.children) > 0


class MCTS:
    def __init__(self, network=None, device="cpu", num_simulations: int = 200,
                 dirichlet_alpha: float = 0.3, dirichlet_eps: float = 0.25,
                 batch_size: int = 16, virtual_loss: float = 1.0, evaluator=None):
        # Evaluation is pluggable: either a local network (LocalEvaluator) or a
        # RemoteEvaluator that round-trips leaf positions to a central batched
        # inference server. `MCTS(network, device=...)` still works.
        if evaluator is None:
            from .evaluator import LocalEvaluator
            evaluator = LocalEvaluator(network, device)
        self.evaluator = evaluator
        self.num_simulations = num_simulations
        self.dirichlet_alpha = dirichlet_alpha
        self.dirichlet_eps = dirichlet_eps
        # Leaf-parallel batching: collect up to `batch_size` leaves (guided away
        # from each other by a temporary virtual loss) and evaluate them in a
        # single network forward pass.
        self.batch_size = max(1, batch_size)
        self.virtual_loss = virtual_loss

    def _mask_probs(self, state: State, probs: np.ndarray):
        """Map network policy (canonical action space) onto a normalized
        distribution over this state's real legal actions; return (probs, mask)."""
        mask = legal_action_mask(state)
        real_probs = np.zeros(ACTION_SIZE, dtype=np.float32)
        for a in np.nonzero(mask)[0]:
            real_probs[a] = probs[encode_action_for_player(state, int(a))]
        total = real_probs.sum()
        if total > 0:
            real_probs /= total
        else:
            real_probs = mask / max(mask.sum(), 1)
        return real_probs, mask

    def _evaluate(self, state: State):
        """Single-state evaluation. Returns (real_probs, value, mask)."""
        return self._evaluate_batch([state])[0]

    def _evaluate_batch(self, states: list):
        """Evaluate several states via the (possibly remote) evaluator.

        Returns a list of (real_probs, value, mask) aligned with `states`."""
        planes = np.stack([encode_state(s) for s in states]).astype(np.float32)
        probs, values = self.evaluator.infer(planes)
        results = []
        for i, state in enumerate(states):
            real_probs, mask = self._mask_probs(state, probs[i])
            results.append((real_probs, float(values[i]), mask))
        return results

    def _collect_leaves(self, root: "Node", remaining: int):
        """Walk the tree to gather a batch of distinct non-terminal leaves
        (with virtual loss applied). Terminal leaves are backed up immediately.

        Returns (leaves, paths, terminal_sims)."""
        leaves, paths = [], []
        pending = set()
        terminal_sims = 0
        target = min(self.batch_size, remaining)
        while len(leaves) < target:
            node = root
            path = [node]
            while node.expanded() and not node.state.is_terminal():
                _, node = self._select_child(node)
                path.append(node)
            leaf = path[-1]
            if leaf.state.is_terminal():
                winner = leaf.state.winner
                value = 1.0 if winner == leaf.player else -1.0
                self._backpropagate(path, value)
                terminal_sims += 1
                if terminal_sims >= remaining:
                    break
                continue
            if id(leaf) in pending:
                # Collision: virtual loss didn't steer us away (e.g. only one
                # leaf left). Evaluate what we have rather than spin.
                break
            self._apply_virtual_loss(path)
            pending.add(id(leaf))
            leaves.append(leaf)
            paths.append(path)
        return leaves, paths, terminal_sims

    def _apply_results(self, results, leaves, paths) -> int:
        for (priors, value, mask), leaf, path in zip(results, leaves, paths):
            self._revert_virtual_loss(path)
            self._expand(leaf, priors, mask)
            self._backpropagate(path, value)
        return len(leaves)

    def run(self, root_state: State, add_noise: bool = True) -> "Node":
        root = Node(root_state.clone())
        priors, _, mask = self._evaluate(root.state)
        self._expand(root, priors, mask)
        if add_noise:
            self._add_dirichlet_noise(root)
        if not root.children:
            return root

        sims_done = 0
        while sims_done < self.num_simulations:
            leaves, paths, term = self._collect_leaves(root, self.num_simulations - sims_done)
            sims_done += term
            if not leaves:
                continue
            results = self._evaluate_batch([lf.state for lf in leaves])
            sims_done += self._apply_results(results, leaves, paths)
        return root

    async def run_async(self, root_state: State, async_eval, add_noise: bool = True) -> "Node":
        """Like run(), but awaits an async (batched) evaluator. Used to run many
        games concurrently in one event loop so their leaves batch together."""
        root = Node(root_state.clone())
        priors, _, mask = (await self._evaluate_batch_async([root.state], async_eval))[0]
        self._expand(root, priors, mask)
        if add_noise:
            self._add_dirichlet_noise(root)
        if not root.children:
            return root

        sims_done = 0
        while sims_done < self.num_simulations:
            leaves, paths, term = self._collect_leaves(root, self.num_simulations - sims_done)
            sims_done += term
            if not leaves:
                continue
            results = await self._evaluate_batch_async([lf.state for lf in leaves], async_eval)
            sims_done += self._apply_results(results, leaves, paths)
        return root

    async def _evaluate_batch_async(self, states: list, async_eval):
        planes = np.stack([encode_state(s) for s in states]).astype(np.float32)
        probs, values = await async_eval.infer(planes)
        results = []
        for i, state in enumerate(states):
            real_probs, mask = self._mask_probs(state, probs[i])
            results.append((real_probs, float(values[i]), mask))
        return results

    def _apply_virtual_loss(self, path: list) -> None:
        # Temporarily discourage re-walking this path within a batch. Selection
        # uses q = -child.value(), so to make a node *less* attractive to its
        # parent we push its value UP (looks good for the node's own mover ->
        # bad for the parent) and add a virtual visit.
        vl = self.virtual_loss
        for node in path:
            node.visit_count += vl
            node.value_sum += vl

    def _revert_virtual_loss(self, path: list) -> None:
        vl = self.virtual_loss
        for node in path:
            node.visit_count -= vl
            node.value_sum -= vl

    def _select_child(self, node: Node):
        best_score, best_action, best_child = -float("inf"), None, None
        sqrt_total = math.sqrt(max(node.visit_count, 1))
        for action, child in node.children.items():
            q = -child.value()  # child stores value for its own player to move
            u = C_PUCT * child.prior * sqrt_total / (1 + child.visit_count)
            score = q + u
            if score > best_score:
                best_score, best_action, best_child = score, action, child
        return best_action, best_child

    def _expand(self, node: Node, priors: np.ndarray, mask: np.ndarray) -> None:
        legal_actions = np.nonzero(mask)[0]
        for action in legal_actions:
            child_state = apply_action(node.state, int(action))
            node.children[int(action)] = Node(child_state, prior=float(priors[action]))

    def _add_dirichlet_noise(self, root: Node) -> None:
        actions = list(root.children.keys())
        if not actions:
            return
        noise = np.random.dirichlet([self.dirichlet_alpha] * len(actions))
        for action, n in zip(actions, noise):
            child = root.children[action]
            child.prior = (1 - self.dirichlet_eps) * child.prior + self.dirichlet_eps * n

    def _backpropagate(self, path: list, leaf_value: float) -> None:
        value = leaf_value
        for node in reversed(path):
            node.visit_count += 1
            node.value_sum += value
            value = -value

    @staticmethod
    def policy_from_visits(root: Node, temperature: float = 1.0) -> np.ndarray:
        policy = np.zeros(ACTION_SIZE, dtype=np.float32)
        actions = list(root.children.keys())
        visits = np.array([root.children[a].visit_count for a in actions], dtype=np.float32)
        if temperature <= 1e-3:
            best = actions[int(np.argmax(visits))]
            policy[best] = 1.0
            return policy
        visits = visits ** (1.0 / temperature)
        visits /= visits.sum()
        for a, p in zip(actions, visits):
            policy[a] = p
        return policy

File: go/mcts/mcts.go
// Package mcts is a Go port of the Python PUCT MCTS, including leaf batching
// with virtual loss (corrected sign) so many leaves evaluate in one forward.
package mcts

import (
	"math"
	"math/rand"

	"github.com/andreiconst/quoridor/engine"
)

// Evaluation is a masked+normalized policy over real actions plus a value from
// the side-to-move's perspective.
type Evaluation struct {
	Policy []float32 // length engine.ActionSize; zero for illegal actions
	Value  float32
}

// Evaluator maps a batch of states to evaluations (one per state).
type Evaluator interface {
	Evaluate(states []*engine.State) []Evaluation
}

// Node is a search-tree node.
type Node struct {
	State      *engine.State
	Prior      float64
	Children   map[int]*Node
	VisitCount float64
	ValueSum   float64
	Player     int
}

func (n *Node) value() float64 {
	if n.VisitCount == 0 {
		return 0
	}
	return n.ValueSum / n.VisitCount
}

func (n *Node) expanded() bool { return len(n.Children) > 0 }

// MCTS holds search configuration.
type MCTS struct {
	Eval           Evaluator
	NumSimulations int
	BatchSize      int
	VirtualLoss    float64
	CPuct          float64
	DirichletAlpha float64
	DirichletEps   float64
	Rng            *rand.Rand
}

// New returns an MCTS with AlphaZero-ish defaults.
func New(eval Evaluator, numSimulations, batchSize int, rng *rand.Rand) *MCTS {
	return &MCTS{
		Eval:           eval,
		NumSimulations: numSimulations,
		BatchSize:      batchSize,
		VirtualLoss:    1.0,
		CPuct:          1.5,
		DirichletAlpha: 0.3,
		DirichletEps:   0.25,
		Rng:            rng,
	}
}

// Run executes the search and returns the populated root.
func (m *MCTS) Run(rootState *engine.State, addNoise bool) *Node {
	root := &Node{State: rootState.Clone(), Player: rootState.Current, Children: map[int]*Node{}}
	rootEval := m.Eval.Evaluate([]*engine.State{root.State})[0]
	m.expand(root, rootEval)
	if addNoise {
		m.addDirichletNoise(root)
	}
	if !root.expanded() {
		return root
	}

	simsDone := 0
	for simsDone < m.NumSimulations {
		leaves, paths, term := m.collectLeaves(root, m.NumSimulations-simsDone)
		simsDone += term
		if len(leaves) == 0 {
			continue
		}
		states := make([]*engine.State, len(leaves))
		for i, lf := range leaves {
			states[i] = lf.State
		}
		results := m.Eval.Evaluate(states)
		for i := range leaves {
			m.revertVirtualLoss(paths[i])
			m.expand(leaves[i], results[i])
			m.backprop(paths[i], float64(results[i].Value))
			simsDone++
		}
	}
	return root
}

func (m *MCTS) collectLeaves(root *Node, remaining int) ([]*Node, [][]*Node, int) {
	var leaves []*Node
	var paths [][]*Node
	pending := map[*Node]bool{}
	terminal := 0
	target := m.BatchSize
	if remaining < target {
		target = remaining
	}
	for len(leaves) < target {
		node := root
		path := []*Node{node}
		for node.expanded() && !node.State.IsTerminal() {
			_, node = m.selectChild(node)
			path = append(path, node)
		}
		leaf := path[len(path)-1]
		if leaf.State.IsTerminal() {
			value := -1.0
			if leaf.State.Winner == leaf.Player {
				value = 1.0
			}
			m.backprop(path, value)
			terminal++
			if terminal >= remaining {
				break
			}
			continue
		}
		if pending[leaf] {
			break // collision: evaluate what we have
		}
		m.applyVirtualLoss(path)
		pending[leaf] = true
		leaves = append(leaves, leaf)
		paths = append(paths, path)
	}
	return leaves, paths, terminal
}

func (m *MCTS) selectChild(node *Node) (int, *Node) {
	best := math.Inf(-1)
	var bestA int
	var bestC *Node
	sqrtTotal := math.Sqrt(math.Max(node.VisitCount, 1))
	for a, child := range node.Children {
		q := -child.value()
		u := m.CPuct * child.Prior * sqrtTotal / (1 + child.VisitCount)
		if score := q + u; score > best {
			best, bestA, bestC = score, a, child
		}
	}
	return bestA, bestC
}

func (m *MCTS) expand(node *Node, e Evaluation) {
	if node.Children == nil {
		node.Children = map[int]*Node{}
	}
	// Masked policy is zero for illegal actions, so nonzero entries are exactly
	// the legal actions (matches the Python expand's legal set).
	for a, p := range e.Policy {
		if p > 0 {
			child := node.State.ApplyAction(a)
			node.Children[a] = &Node{State: child, Prior: float64(p), Player: child.Current}
		}
	}
}

func (m *MCTS) backprop(path []*Node, leafValue float64) {
	value := leafValue
	for i := len(path) - 1; i >= 0; i-- {
		path[i].VisitCount++
		path[i].ValueSum += value
		value = -value
	}
}

// applyVirtualLoss discourages re-walking a path within a batch. Selection uses
// q = -child.value(), so to make a node less attractive to its parent we push
// its value UP (good for the node's own mover) and add a virtual visit.
func (m *MCTS) applyVirtualLoss(path []*Node) {
	for _, n := range path {
		n.VisitCount += m.VirtualLoss
		n.ValueSum += m.VirtualLoss
	}
}

func (m *MCTS) revertVirtualLoss(path []*Node) {
	for _, n := range path {
		n.VisitCount -= m.VirtualLoss
		n.ValueSum -= m.VirtualLoss
	}
}

func (m *MCTS) addDirichletNoise(root *Node) {
	actions := make([]int, 0, len(root.Children))
	for a := range root.Children {
		actions = append(actions, a)
	}
	if len(actions) == 0 {
		return
	}
	noise := dirichlet(m.Rng, m.DirichletAlpha, len(actions))
	for i, a := range actions {
		c := root.Children[a]
		c.Prior = (1-m.DirichletEps)*c.Prior + m.DirichletEps*noise[i]
	}
}

// PolicyFromVisits returns a length-ActionSize distribution from visit counts.
func PolicyFromVisits(root *Node, temperature float64) []float32 {
	policy := make([]float32, engine.ActionSize)
	if temperature <= 1e-3 {
		bestA, bestV := -1, -1.0
		for a, c := range root.Children {
			if c.VisitCount > bestV {
				bestV, bestA = c.VisitCount, a
			}
		}
		if bestA >= 0 {
			policy[bestA] = 1
		}
		return policy
	}
	var sum float64
	weights := map[int]float64{}
	for a, c := range root.Children {
		w := math.Pow(c.VisitCount, 1.0/temperature)
		weights[a] = w
		sum += w
	}
	if sum > 0 {
		for a, w := range weights {
			policy[a] = float32(w / sum)
		}
	}
	return policy
}

// --- Dirichlet via gamma sampling ---

func dirichlet(rng *rand.Rand, alpha float64, n int) []float64 {
	out := make([]float64, n)
	var sum float64
	for i := range out {
		out[i] = gammaSample(rng, alpha)
		sum += out[i]
	}
	if sum > 0 {
		for i := range out {
			out[i] /= sum
		}
	}
	return out
}

// gammaSample draws from Gamma(alpha, 1) (Marsaglia-Tsang, with boost for a<1).
func gammaSample(rng *rand.Rand, alpha float64) float64 {
	if alpha < 1 {
		u := rng.Float64()
		return gammaSample(rng, alpha+1) * math.Pow(u, 1.0/alpha)
	}
	d := alpha - 1.0/3.0
	c := 1.0 / math.Sqrt(9*d)
	for {
		x := rng.NormFloat64()
		v := 1 + c*x
		if v <= 0 {
			continue
		}
		v = v * v * v
		u := rng.Float64()
		if u < 1-0.0331*x*x*x*x {
			return d * v
		}
		if math.Log(u) < 0.5*x*x+d*(1-v+math.Log(v)) {
			return d * v
		}
	}
}
